codeguru_profiler_agent/model/profile.py

```diff
@@ -56,11 +58,43 @@ def get_active_millis_since_start(self):
         This returns the total "active" wall clock time since start. In AWS lambda the process can be frozen, the
         time while we are frozen should not be counted in here. In an EC2 or other type of host it is simply the wall
         clock time since start.
-        If the "end" time has been set, we give the active time between start and end otherwise between start and now
+
+        Previously self.end was used to calculate active_millis_since_start but self.end is updated when a sample is added
+        so in rare cases where a sample is collected before the last pause time then we might add additional pause time
+        which can lead to incorrect calculation of active time. In worst cases, it can even lead to negative values depending
+        on the pause time.
+
+        Below is an example indicating the potential error of considering self.end in the calculation.
+        ------------------------------------------------------------------
+        |          |        |    |         |    |     |       |     ||||||
+        S          P        R    P         R    SE    P       R     REPORT
+        s1_________|___p1___|____|____p2___|____e1____|___p3__|__________|
+
+        S - Start
+        P - Pause
+        R - Resume
+        SE - self.end (last sample)
+        REPORT - Calculating active time.
+
+        If we consider self.end which is e1 in above case then active time would be e1-s1-(p1+p2+p3). But pause p3 is after e1
+        so that leads to incorrect calculation of active time.
+
+        Ideally we would want to set profile_end to be last sample time and subtract pause times only before that but it requires
+        additional work in maintaining pause time which isn't worth as it makes the logic complex with very little gain.
+
+        So we are setting it to current time and in some corner cases to last_pause time.
+
+        Additional note: last_pause is set to none as part of resume and _paused_ms is added as part of resume. So we won't
```

Comment:
Actually it can still happen if we resume while calculating the active millis, we could have the following happening:
main thread                                                                                 profiler thread
`self._paused_ms += self.last_resume - self.last_pause`           
                                                                                                      `if self.last_pause is not None:`
                                                                                                      `    end = self.last_pause`
`self.last_pause = None`
                                                                                                      `end - self.start - self._paused_ms` <- too small
                                                                                                      

Maybe in the resume we should set `self.last_pause` to None before we update `self._paused_ms`

Reply:
Yeah, if it happens concurrently then we might end up in a buggy state. But I am not sure if it is possible. 

I don't think reordering or multiple assignment would solve it completely but takes care of most of the scenario so I have updated it. In case we come across such a bug then we can look into solving it properly.